File: packages/server/src/services/model-providers/provider-key-service.ts
// packages/server/src/services/model-providers/provider-key-service.ts
import { providerKeyStorage } from '@/utils/storage/provider-key-storage' // ADDED
import {
  CreateProviderKeyInputSchema,
  ProviderKey,
  ProviderKeySchema,
  UpdateProviderKeyInput
} from 'shared/src/schemas/provider-key.schemas'
import { ApiError } from 'shared'
import { z } from '@hono/zod-openapi'
import { normalizeToUnixMs } from '@/utils/parse-timestamp'

// The mapDbRowToProviderKey function is no longer needed as we store objects directly
// that should conform to the ProviderKey schema.

export type CreateProviderKeyInput = z.infer<typeof CreateProviderKeyInputSchema>

/**
 * Returns an object of functions to create, list, update, and delete provider keys,
 * using JSON file storage.
 */
export function createProviderKeyService() {
  async function createKey(data: CreateProviderKeyInput): Promise<ProviderKey> {
    const allKeys = await providerKeyStorage.readProviderKeys()
    const now = normalizeToUnixMs(new Date())
    const id = providerKeyStorage.generateId()

    const newKeyData: ProviderKey = {
      id,
      provider: data.provider,
      key: data.key, // Stored in plaintext initially
      created: now,
      updated: now,
      // Add any other fields from ProviderKeySchema with defaults if necessary
      // e.g., metadata: data.metadata ?? null, if metadata was part of your schema
    }

    // Validate the new key data against the schema before saving
    const parseResult = ProviderKeySchema.safeParse(newKeyData)
    if (!parseResult.success) {
      console.error(`Validation failed for new provider key data: ${parseResult.error.message}`, {
        rawData: data,
        constructedData: newKeyData,
        error: parseResult.error.flatten()
      })
      throw new ApiError(500, 'Internal validation error creating provider key.', 'PROVIDER_KEY_VALIDATION_ERROR', parseResult.error.flatten())
    }

    const validatedNewKey = parseResult.data;

    if (allKeys[validatedNewKey.id]) {
      // Extremely unlikely with UUIDs but a good safeguard
      throw new ApiError(500, `Provider key ID conflict for ${validatedNewKey.id}`, 'PROVIDER_KEY_ID_CONFLICT')
    }

    allKeys[validatedNewKey.id] = validatedNewKey
    await providerKeyStorage.writeProviderKeys(allKeys)
    return validatedNewKey
  }

  async function listKeys(): Promise<ProviderKey[]> {
    const allKeys = await providerKeyStorage.readProviderKeys()
    const keyList = Object.values(allKeys)

    // Sort by provider, then by created descending (as in original SQL)
    keyList.sort((a, b) => {
      if (a.provider < b.provider) return -1
      if (a.provider > b.provider) return 1
      // Assuming created are valid ISO strings, direct string comparison for descending order
      if (a.created > b.created) return -1
      if (a.created < b.created) return 1
      return 0
    })
    return keyList
  }

  async function getKeyById(id: number): Promise<ProviderKey | null> {
    const allKeys = await providerKeyStorage.readProviderKeys()
    const foundKeyData = allKeys[id]

    if (!foundKeyData) {
      return null
    }
    // Data should already be validated by readValidatedJson via providerKeyStorage.
    // If an extra check is desired, uncomment:
    /*
    const parseResult = ProviderKeySchema.safeParse(foundKeyData)
    if (!parseResult.success) {
      console.error(`Failed to parse provider key data for ID ${id} from storage. Data integrity issue.`, {
        id,
        foundData: foundKeyData,
        error: parseResult.error.flatten()
      })
      throw new ApiError(
        500,
        `Failed to parse provider key data for ID ${id}. Data integrity issue.`,
        'PROVIDER_KEY_PARSE_FAILED_ON_READ',
        { id, foundData: foundKeyData, error: parseResult.error.flatten() }
      )
    }
    return parseResult.data
    */
    return foundKeyData
  }

  async function updateKey(id: number, data: UpdateProviderKeyInput): Promise<ProviderKey> {
    const allKeys = await providerKeyStorage.readProviderKeys()
    const existingKey = allKeys[id]

    if (!existingKey) {
      throw new ApiError(404, `Provider key with ID ${id} not found for update.`, 'PROVIDER_KEY_NOT_FOUND_FOR_UPDATE')
    }

    const updatedKeyData: ProviderKey = {
      ...existingKey,
      provider: data.provider ?? existingKey.provider,
      key: data.key ?? existingKey.key, // Stored in plaintext initially
      updated: normalizeToUnixMs(new Date())
      // any other updatable fields
    }

    const parseResult = ProviderKeySchema.safeParse(updatedKeyData)
    if (!parseResult.success) {
      console.error(`Validation failed updating provider key ${id}: ${parseResult.error.message}`, {
        id,
        updatePayload: data,
        mergedData: updatedKeyData,
        error: parseResult.error.flatten()
      })
      throw new ApiError(500, `Internal validation error updating provider key.`, 'PROVIDER_KEY_UPDATE_VALIDATION_ERROR', parseResult.error.flatten())
    }

    const validatedUpdatedKey = parseResult.data;
    allKeys[id] = validatedUpdatedKey
    await providerKeyStorage.writeProviderKeys(allKeys)
    return validatedUpdatedKey
  }

  async function deleteKey(id: number): Promise<boolean> {
    const allKeys = await providerKeyStorage.readProviderKeys()
    if (!allKeys[id]) {
      return false // Key not found, nothing to delete
    }

    delete allKeys[id]
    await providerKeyStorage.writeProviderKeys(allKeys)
    return true
  }

  return {
    createKey,
    listKeys,
    getKeyById,
    updateKey,
    deleteKey
  }
}

export const providerKeyService = createProviderKeyService()

File: packages/python_backend/app/services/provider_key_service.py
from datetime import datetime, timezone
from typing import List, Optional
from fastapi import HTTPException
from pydantic import ValidationError

from app.schemas.provider_key_schemas import (
    ProviderKey,
    CreateProviderKeyBody,
    UpdateProviderKeyBody,
    ProviderKeyListResponse
)
from app.utils.storage.provider_key_storage import provider_key_storage_util, ProviderKeysStorage

class ProviderKeyService:
    async def create_key(self, data: CreateProviderKeyBody) -> ProviderKey:
        all_keys = await provider_key_storage_util.read_provider_keys()
        now_ms = provider_key_storage_util.generate_id()
        key_id = provider_key_storage_util.generate_id()

        new_key_data_dict = {
            "id": key_id,
            "provider": data.provider,
            "key": data.key,
            "created": now_ms,
            "updated": now_ms,
        }
        try:
            new_key = ProviderKey.model_validate(new_key_data_dict)
        except ValidationError as e:
            raise HTTPException(status_code=500, detail={"message": "Internal validation error creating provider key.", "code": "PROVIDER_KEY_VALIDATION_ERROR", "errors": e.errors()})

        if new_key.id in all_keys:
            raise HTTPException(status_code=500, detail={"message": f"Provider key ID conflict for {new_key.id}", "code": "PROVIDER_KEY_ID_CONFLICT"})

        all_keys[new_key.id] = new_key
        await provider_key_storage_util.write_provider_keys(all_keys)
        return new_key

    async def list_keys(self) -> List[ProviderKeyListResponse.ProviderKeyListItem]:
        all_keys = await provider_key_storage_util.read_provider_keys()
        key_list = list(all_keys.values())
        key_list.sort(key=lambda k: k.created, reverse=True)
        key_list.sort(key=lambda k: k.provider)
        
        list_items = []
        for key in key_list:
            list_item = ProviderKeyListResponse.ProviderKeyListItem(
                id=key.id,
                provider=key.provider,
                created=key.created,
                updated=key.updated
            )
            list_items.append(list_item)
        return list_items

    async def get_key_by_id(self, key_id: int) -> Optional[ProviderKey]:
        all_keys = await provider_key_storage_util.read_provider_keys()
        return all_keys.get(key_id)

    async def get_keys_by_provider(self, provider_name: str) -> List[ProviderKey]:
        all_keys = await provider_key_storage_util.read_provider_keys()
        provider_specific_keys = [key for key in all_keys.values() if key.provider == provider_name]
        provider_specific_keys.sort(key=lambda k: k.created, reverse=True)
        return provider_specific_keys

    async def list_all_key_details(self) -> List[ProviderKey]:
        all_keys = await provider_key_storage_util.read_provider_keys()
        key_list = list(all_keys.values())
        key_list.sort(key=lambda k: k.created, reverse=True)
        key_list.sort(key=lambda k: k.provider)
        return key_list

    async def update_key(self, key_id: int, data: UpdateProviderKeyBody) -> ProviderKey:
        all_keys = await provider_key_storage_util.read_provider_keys()
        existing_key = all_keys.get(key_id)
        if not existing_key:
            raise HTTPException(status_code=404, detail={"message": f"Provider key with ID {key_id} not found for update.", "code": "PROVIDER_KEY_NOT_FOUND_FOR_UPDATE"})

        update_data_dict = data.model_dump(exclude_unset=True)
        updated_key_data = existing_key.model_copy(update=update_data_dict)
        updated_key_data.updated = provider_key_storage_util.generate_id()
        try:
            ProviderKey.model_validate(updated_key_data.model_dump())
        except ValidationError as e:
            raise HTTPException(status_code=500, detail={"message": "Internal validation error updating provider key.", "code": "PROVIDER_KEY_UPDATE_VALIDATION_ERROR", "errors": e.errors()})

        all_keys[key_id] = updated_key_data
        await provider_key_storage_util.write_provider_keys(all_keys)
        return updated_key_data

    async def delete_key(self, key_id: int) -> bool:
        all_keys = await provider_key_storage_util.read_provider_keys()
        if key_id not in all_keys:
            return False
        del all_keys[key_id]
        await provider_key_storage_util.write_provider_keys(all_keys)
        return True

# Instantiate the service for use in routes
provider_key_service = ProviderKeyService()
